LanguageExt.Sys/Live/Runtime.cs

```diff
@@ -133,6 +136,13 @@ namespace LanguageExt.Sys.Live
         /// <returns>Operating-system environment environment</returns>
         public Eff<Runtime, Traits.EnvironmentIO> EnvironmentEff =>
             SuccessEff(Sys.Live.EnvironmentIO.Default);
+        
+        /// <summary>
+        /// Access the random synchronous effect environment
+        /// </summary>
+        /// <returns>Random synchronous effect environment</returns>
+        public Eff<Runtime, Traits.RandomIO> RandomEff =>
+            SuccessEff<Runtime, Traits.RandomIO>(new Sys.Live.RandomIO(new Random()));
```

Comment:
Creating a new random generator here, means you will use a new random generator for each line in the following example:

```csharp
from x in Random<RT>.nextInt()
from y in Random<RT>.nextInt()
from z in Random<RT>.nextInt()
select x + y + z
```

When I want to use the same random generator I can't use the static `Runtime<RT>` helper methods.
That would be sad.

I don't have a nice solution, but here is an idea:

The live runtime holds the `Random` instance and the seed in its `RuntimeEnv`.
The live `RandomIO` synchronizes the access to the random generator from multiple threads. (e.g. https://learn.microsoft.com/en-us/dotnet/api/system.random?view=netstandard-2.1#the-systemrandom-class-and-thread-safety).
Similar to the `LocalCancel` property of the `HasCancel<RT>` trait, we add a `LocalRandom` property which allows for a new random generator in case I don't want to use the runtime global one.
Maybe `LocalRandom` should be a method, where I can use a different seed.

Reply:
Thanks for review!

I agree, will make some changes there,

* I expect 1 random instance per runtime
* A local random with optional seed would be a great feature

Reply:
@timmi-on-rails have made the changes so that Random is scoped to the runtime instance.

Not 100% on-board with the recommended approach outlined above, need to talk through the benefits of a singleton source of randomness, as apposed to a random instance scoped to the runtime.

Reply:
I didn't mean to create a singleton.
One Random instance per runtime was my intention as well.
Yet, you need to address thread- safety.
I will sight your changes later.

Reply:
All good, I  understand.

Might start by creating a test that corrupts the Random instance and try a few synchronization approaches. 

* The recommended approach with locks
* The Random instance per thread local approach
* Some sort of interlock variation. Not going lie, low level concurrency is not my strongest skill

The goals,

* No corrupt Random, users should not get streams of zeros
* Support for user seeds, and local seeds.
* Performance is not compromised

Reply:
@timmi-on-rails random IO is now fully thread safe and the live instance is singleton scoped to each thread that accesses it (using thread local).

Should provide for the best of both worlds.

Deterministic random generators for tests, correct unbiased and fast random generation live.

User provided seeds for the Random generator are supported, the live instance memoizes the thread local instance for all unique seeds used.

The risk here is that the user leaks memory by randomly generating seeds, If that is done, the user is on their own.

The case would be a misuse of seeds and would be pointless IMHO.

If you are happy you can resolve it.

Reply:
Why did you go for a static Random instance per thread?
I would prefer runtime-scoped Random instances for better isolation.

Reply:
Because it generates more uniform sequences and is lock free.

If you have a 100% pure application it will effectively be a singleton, as there is a single entry point. However it's not thread safe so locks are required.

But consider many hybrid codebases with many entry points, the runtime does not live very long. So might cause sequence issues and still requires locks.

With thread scoped singleton random instances it complies with the recommended approach outlined by MS, generates the most uniform sequences under any codebase configuration and is lock free.

Happy to consider counter arguments.

Reply:
The point about short-lived runtimes convinced me, e.g. runtime per request in a webserver.

(BTW: Thanks and I will answer your comment on my EF Core thread. Currently too busy.)